File: Editorial/0101-0200/0106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/C/1_Recursion.c
struct TreeNode* buildTree(int* inorder, int inorderSize, int* postorder,
                           int postorderSize) {
    if (inorderSize == 0) {
        return NULL;
    }
    int i;
    for (i = 0; i < inorderSize; i++) {
        if (inorder[i] == postorder[postorderSize - 1]) {
            break;
        }
    }
    struct TreeNode* root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->val = postorder[postorderSize - 1];
    root->left = buildTree(inorder, i, postorder, i);
    root->right = buildTree(inorder + i + 1, inorderSize - i - 1, postorder + i,
                            postorderSize - i - 1);
    return root;
}

File: Editorial/0101-0200/0106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/CSharp/1_Recursion.cs
public class Solution {
    int post_idx;
    int[] postorder;
    int[] inorder;
    Dictionary<int, int> idx_map = new Dictionary<int, int>();

    public TreeNode Helper(int in_left, int in_right) {
        if (in_left > in_right)
            return null;
        int root_val = postorder[post_idx];
        TreeNode root = new TreeNode(root_val);
        int index = idx_map[root_val];
        post_idx--;
        root.right = Helper(index + 1, in_right);
        root.left = Helper(in_left, index - 1);
        return root;
    }

    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        this.postorder = postorder;
        this.inorder = inorder;
        post_idx = postorder.Length - 1;
        for (int idx = 0; idx < inorder.Length; idx++)
            idx_map[inorder[idx]] = idx;
        return Helper(0, inorder.Length - 1);
    }
}
